### DISCUSSION TITLE
MQTT topics and messages

### DISCUSSION THREAD

**Discussion by paddywwoof** [helgeerbe/picframe#19] [Show and tell]
I will put these comments in a new discussion to make them easier to reference (hopefully this has now settled down and won't change too much):
00. There are two styles of topic depending on how communication is being used by homeassistant. In the list below I will give the topic followed by possible messages. First switch or button style, then value messages

0. on the basis that you have defined ```  device_id: 'picframe'``` in configuration.yaml
1. display on/off: ```homeassistant/switch/picframe_display/set```, ```ON,OFF```
2. paused on/off: ```homeassistant/switch/picframe_paused/set```, ```ON,OFF```
3. shuffle on/off: ```homeassistant/switch/picframe_shuffle/set```, ```ON,OFF```
4.  back: ```homeassistant/switch/picframe_back/set```, ```ON```
5.  next: ```homeassistant/switch/picframe_next/set```, ```ON```
6.  name text on/off: ```homeassistant/switch/picframe_name_toggle/set```, ```ON,OFF```
7.  date text on/off: ```homeassistant/switch/picframe_date_toggle/set```, ```ON,OFF```
8.  location text on/off: ```homeassistant/switch/picframe_location_toggle/set```, ```ON,OFF```
9.  directory text on/off: ```homeassistant/switch/picframe_directory_toggle/set```, ```ON,OFF```
10.  title text on/off: ```homeassistant/switch/picframe_title_toggle/set```, ```ON,OFF```
11.  caption text on/off: ```homeassistant/switch/picframe_caption_toggle/set```, ```ON,OFF```
12.  all text off: ```homeassistant/switch/picframe_text_off/set```, ```ON``` # yes that is confusing!
13.  text refresh: ```homeassistant/switch/picframe_text_refresh/set```, ```ON```
14.  delete: ```homeassistant/switch/picframe_delete/set```, ```ON```

values:
1.  directory: ```picframe/directory```, ```string matching directory name```
2. date_from: ```picframe/date_from```, ```string YYYY/MM/DD or number representing unix timestamp or zero length string to remove constraint```
3. date_to: ```picframe/date_to```, ```string YYYY/MM/DD or number as above```
4. fade_time: ```picframe/fade_time```, ```number```
5. time_delay: ```picframe/time_delay```, ```number```
6. brightness: ```picframe/brightness```, ```float 0.0 to 1.0```
7. location_filter: ```picframe/location_filter```, ```string matching any part of geolocation description```
8. tags_filter: ```picframe/tags_filter```, ```string matching tags```
9. stop: ```picframe/stop```, message not used - just stops

**Comment by helgeerbe** [comment]
I have to check, but I think 

- date_from: picframe/date_from, float  timestamp
- date_to: picframe/date_to, float  timestamp

**Reply by paddywwoof** [reply]
Yes, it can accept both. I will edit it.

**Reply by helgeerbe** [reply]
Paddy, I just started the wiki pages. I will move this discussion to the documentation. Good thing is, that the wiki is just an other repro. So we could use git to work offline on it. Just started with the homepage. Next steps: create table of content. So I can fill the chapters, step by step.

**Reply by paddywwoof** [reply]
Excellent. Once you've got an outline we can work on one part at a time each - share the work out. Whatever we do Wolfgang's version will be the **definitive** version!

**Reply by jgodfrey** [reply]
> Excellent. Once you've got an outline we can work on one part at a time each - share the work out. Whatever we do Wolfgang's version will be the definitive version!

Agreed - happy to help when / where I can.

**Reply by helgeerbe** [reply]
Paddy, Jeff, I invited you as collaborators. So you can pull the wiki, create branches and push them. Only main is visible on github, but locally you can work normally. I made a layout structure in the side bar and created a Getting-Started page. Let me know, if the structure makes sense for you.

**Reply by paddywwoof** [reply]
Helge, looks good so far. I will clone it too - not worked with the github wiki system before but it looks straightforward. Hopefully the code is reaching a point of stability - just the 'automatic' switch between mat and cropping, filling depending on aspect ratio we talked about.

**Reply by jgodfrey** [reply]
>  just the 'automatic' switch between mat and cropping, filling depending on aspect ratio we talked about.

I've been running an experimental version this afternoon and, I must say, I like the idea. I'm simply doing this (in `tex_load()` ATM...

```python
  disp_aspect = self.__display.width / self.__display.height
  image_aspect = pics[0].width / pics[0].height
  diff_aspect = abs(disp_aspect - image_aspect)
  if self.__mat_images and diff_aspect > 0.25:
      if not pics[1]:
          im = self.__matter.mat_image((im,))
      else:
          im = self.__matter.mat_image((im, im2))
  else:
      if pics[1]: #i.e portrait pair
          im = self.__create_image_pair(im, im2)
```

Obviously, very hacky right now, but seems to do what I'd expect. With some cleanup, and some appropriate configuration backing, it probably doesn't require much more. Though, there's still the bigger issue of making the various display methods work more cohesively together.

**Reply by paddywwoof** [reply]
Jeff, that seems good and is probably going to be sufficient. Looking at the code for blur_edges it does a similar thing with the ```diff_aspect``` (but calls it ```wh_rat```!) and only attempts to do the blurring if it's out by more than 1%.

If the image(s) have been through the matting process then the resultant ```im``` will be exactly the same aspect ratio as the display so the blurring won't happen. The other edge filling systems are zero cost done by the shader.
```python
            if self.__blur_edges and size is not None:
                wh_rat = (size[0] * im.size[1]) / (size[1] * im.size[0])
                if abs(wh_rat - 1.0) > 0.01: # make a blurred background
                    ...
```
If your code recalculated ```diff_aspect``` after doing ```__matter.mat_image``` or ```__create_image_pair``` then that could be used instead of the ```wh_rat``` for deciding the blurring later (more for consistency than to save on processing).

Out of curiosity I tried making the k-means grouping five dimensional by adding the pixel coordinates as well as the rgb value. It occurred to me that maybe a spread-out alternating colour might be less noticeable than a solid block of one colour concentrated in one place. I'm not sure that it makes any difference but I left the code in (commented out).

Another thing someone [just asked](https://github.com/pi3d/pi3d_demos/issues/62) relates to writing back to the db. [I'm not sure that's the full solution as ideally the image would be rotated then saved over itself - but I think that would loose existing exif info unless we first read it with the 'raw' ```exif``` argument and save it with the same argument - I've not tried doing that and I think it only works for jpeg] Anyway, probably a back burner job, but something to be aware of relative to your (much) earlier comment about saving info, modifying db and db schemas.

Paddy

PS
```python
im = Image.open('path_to_image.jpg')
exif = im.info['exif']
im = im.transpose(Image.ROTATE_270)
im.save('path_to_image.jpg', exif=exif) # normally this image will be 200k if it was originally 1.3M with no apparent loss of quality. Useful to know if you want to increase your storage capacity x6!!!
```

**Reply by jgodfrey** [reply]
> ```python
> im.save('path_to_image.jpg', exif=exif) # normally this image will be 200k if it was originally 1.3M with no apparent loss of quality. Useful to know if you want to increase your storage capacity x6!!!
> ```

Hmmm... That has to be due to 1) the likely lower resolution and 2) the default JPEG quality of the JPG writer.

Yeah, poking around in the PIL docs, I see that the JPEG writer defaults to 75% quality, which seems a bit low to me.

Also, I'm not sure I like the whole idea of allowing the picture frame software to *modify* the images it's given to display. That seems like a slippery slope and would be wholly unexpected for most Users I think (though, I assume it'd just be a configurable thing).

Really, I guess its a question of perceived ownership. That is, is it OK for the picture frame to _own_ (and therefore, do what it wants with) the pictures it has access to? Or, are the pictures owned by the User, with the frame only being provided access for purposes of displaying them?

On the surface, I'm in that second camp - those are *my* images and I wouldn't expect the frame to change them on its own. 

At the very least, we'd need to be very careful that a User understands this mode (should it exist). It'd be really bad for someone to aim the frame at some network location with their *golden* copy of their original images, only to find out they've all been severely and permanently downgraded by the frame.

**** EDIT ****

To put it bluntly, this concerns me a bit. I'm confident that no matter how careful we are, at some point, a User will irreparably modify a set of original images and have no recovery plan.

**Reply by jgodfrey** [reply]
Related to the image rotation and resave topic... I guess I don't really understand the original motivation in the above-linked thread. It seems like Rich has a set of images that he's edited (in PS or elsewhere) and then exported as JPG - which has removed the EXIF data from the exported file (which is somewhat typical).

So, now, the picture frame app doesn't have any info to base an auto-rotation off of, so the image is just displayed "as-is". If that "as-is" orientation is wrong, that implies that it was exported at an incorrect orientation. Why was it not corrected in PS prior to the export?

If I have a "sideways" image with no EXIF data to base a correction on, the frame software seems a bit late in the workflow to try and correct that. To me, the solution is to either 1) include the EXIF data so the frame can correct it or 2) properly orient the exported image so the frame doesn't need to correct it.

Maybe I'm missing the whole point?

**Reply by helgeerbe** [reply]
Jeff, we are on the same page. We shouldn't manipulate any images. There are a lot of tools which are capable to handle this. Some people might just mount there NAS directory. And it seems to me to be a bad idea to change anything without asking for permission.

**Reply by paddywwoof** [reply]
Yes obviously there is an increase in compression so there must be a loss in quality. However for 95% of photographs the difference is imperceptible... for the the other 5% that have large areas of *almost* the same colour the drop in quality to 75% does sometimes produce visible compression artifacts! As it doesn't seem possible to find the quality level the image was originally saved at this process is probably fraught with danger as you say.

I also can't really see why you would have pictures that were rotated but didn't have exif info. Maybe they are are slides that have been scanned, but why would you want to use the picture frame as a photo editor when it's trivially easy to run them through any number of image editors (or most viewers that come with OS) and correct the orientation.

So we should explicitly exclude image editing from the scope of picture_frame but it still might be relevant to store info back onto the db at some stage.

Paddy

**Reply by helgeerbe** [reply]
Absolute!  My family asked, if it might be possible to do the image paring only with images, that have related colors. Or to do the pairing selection manual on the browser interface. Some of my images don't have a title. For a little tweak it would be nice, to change that on the fly.  But you have to be aware, that the info is lost, if you drop the database. So have in mind. If we start this, we have to deal with updates and backups. Because, the database is then not a simple cache any more.

**Reply by paddywwoof** [reply]
I suppose the colour could be calculated using the function used in the mat_image file but it's a non-trivial exercise to find the optimal pairing of nearest points in 3D space - it wouldn't have to be very accurate, but even so I suspect at least O(n<sup>2</sup>). However the initial entry onto the db is done without a full load of the pixels as would be needed to determine the representative colour. At the moment the mat colour is based on only the left image, as is the text etc - that's probably a more urgent, and also quite hard, problem to fix. In the short term we should probably just make a note of these things and mop them up later (the github issue system is actually very good for this: assigning priority, responsibility etc)

I've also just pushed a final (famous last words!) fix of a very annoying bug cause by ```L``` or ```1``` mode images with the mat calculation and rectangle drawing. Took ages to track down - I had to remove the try: except: blocks in order to see details of the errors and even then it was far from clear.

Next job is probably to do some installs on RPi3 and RPi4 clean SD cards and see what we've forgotten.

Paddy

**Reply by jgodfrey** [reply]
> At the moment the mat colour is based on only the left image, as is the text etc - that's probably a more urgent, and also quite hard, problem to fix. 

When I did the image color selection initially, I considered combining all (1 or 2) images together before making the color selection. While that'd be trivial to do, I'm not sure whether it'd make (visual) sense in the end. That said, I never tried it. 

Regarding the text overlay info. I've been meaning to ask about that. Really, we have all of the info for both images, and it should be trivial to access it. I guess the bigger issue is in how / where to display it for this case. I have a small-format digital frame here that's able to present portrait pairs. With the text overlay on, it simply presents 2 lines of text (top line is left image, bottom line is right image). Alternatively, maybe we could present the text on 2 halves of the screen, to sort of mimic the image layout. Minimally, I'd guess the text would need to be smaller than (the default) is today. Is there other issues here too that I've missed?

> In the short term we should probably just make a note of these things and mop them up later (the github issue system is actually very good for this: assigning priority, responsibility etc)

Yep. I agree.

> I've also just pushed a final (famous last words!) fix of a very annoying bug cause by `L` or `1` mode images with the mat calculation and rectangle drawing. Took ages to track down - I had to remove the try: except: blocks in order to see details of the errors and even then it was far from clear.

Good find. I haven't tested with any grayscale images. Though, I will say that having all of the errors gobbled up by try/except blocks was driving me crazy too during some of my work. That's why I added the `raise` that's currently at the bottom of `tex_load()`.  In my case, that was enough for me to see what was happening...

> Next job is probably to do some installs on RPi3 and RPi4 clean SD cards and see what we've forgotten.

Yeah, that'd be great. I'm about to order a few more RPi4's. Until then, I only have one.

Somewhat related... I'd like to check out your http work, but haven't even begun to look into it. Is there a set of instructions buried in a thread somewhere that might help me get started?

**Reply by paddywwoof** [reply]
Jeff, I suppose the alternatives for the text are a) keep one full width text block and append the info for the second image below the first (maybe with a ```LEFT:``` and ```RIGHT:``` b) have two text blocks and make the character wrap width half as big. Obviously both approaches would need extra code to determine if it was a pair showing in order to generate the second chunk of text. More tedious to do than difficult.

There probably is something in one of the threads about setting up http. It should just work if you set the config use_http: True and make sure the path leads to directory where the files were copied. If you browse to ```192.168.0.987:9000``` (whatever the RPi IP is and whatever port number you use) then it should serve the index.html page. There is also a user option ```192.168.0.987:9000/current_image``` that will render whatever is showing on the screen to your browser.

You shouldn't need to send GET arguments directly as the js does that for you but for reference: The Controller methods that have @setter decorators can be called by ```?brightness=0.5``` but the other methods need to have the arguments passed as json of kwargs i.e. ```next={}``` or ```set_show_text={"txt_key":"name", "val"="ON"}

As an exercise I should write this up and put it into Helge's wiki!

**Comment by jgodfrey** [comment]
> Jeff, that seems good and is probably going to be sufficient. Looking at the code for blur_edges it does a similar thing with the diff_aspect (but calls it wh_rat!) and only attempts to do the blurring if it's out by more than 1%.

Regarding the above discussion about aspect ratio comparison and auto-matting images that are off by more than some defined percentage... We should agree on a method to calculate the difference between 2 aspect ratios as our methods currently differ.

In the case where the image and the screen aspect both match, our methods agree on that. However, when they don't match, the resulting values are (sometimes considerably) different.

I'm not sure there's any "standard" way to calculate such a difference (since it's 2D data represented by a single scalar value) - and probably it doesn't matter too much, but they should definitely be the same. 

Any thoughts on which is better, or if there's a better way than either of the existing methods?

**Reply by paddywwoof** [reply]
Yes, I've not really thought about it, and +/- 1% won't matter, probably. Logically we're interested in the proportion of area that would be left either top and bottom or either side so that simplifies to
```python
if disp_aspect > image_aspect:
    edge_proportion = 1 - image_aspect / disp_aspect
else:
    edge_proportion = 1 - disp_aspect / image_aspect
```

**Comment by helgeerbe** [comment]
Copy it to the wiki https://github.com/helgeerbe/picframe/wiki/Control-Interfaces
